Add new file for a new sort alg

alerrian opened the pull request:
I added a new file because I wanted to use a class structure for when I add more sorting algorithms later.

I have a little bit of a background with computer science(from college) and was rustling through my old notes when I ran across some of my stuff from a class I took about a year and a half ago called "Data structures and Algorithms" and wanted to see if I would be able to implement code I had there in to Ruby since it was originally in Java. I probably wrote more comments than I wrote actual code, but I tried to describe each line to the best of my ability. Next steps are to write algorithms for merge sorts and quick sorts(which is actually what the .sort array method is using behind the scenes). 

I will be working on some refactoring in the algorithm_sort.rb file since I think it might be better served by taking it apart and using a class structure to work on it. This way, we can add methods or remove them as necessary without effecting anything else. This will take more time and I wanted to get a merge up here to make sure I had the process right.

alerrian committed:
Add bubble_sort method to SortingAlg class.

alerrian committed:
Add new description

iEv0lv3 commented:
This is a great Bubble Sort! I really like this contribution, and I have some ideas about how we can make an integration with the other code work 💻It's super late, so I'll explain my code thoughts a bit more coherently tomorrow 😪 

I see you created your own branch `SAA`, and initiated your pull request into the `dev` branch; that's perfect, well done and thank you! 👍

I posted this suggestion on Pull Request #5 and I thought you mind find it helpful:

To check that both `master` and `dev` are up to date, and you can pull them to your local clone. This means you'll have the most recent updates on your local clone of the repo, which reduces any merge conflicts when pull requests are made. Here is a good outline of the process from StackOverflow (edited by me to use correct names for this project): 

(`your_branch` = whatever the name of the branch is you've been working on)

1. Go to `dev` branch with: `git checkout dev`
2. Update your local `dev` branch with: `git pull origin dev`
3. Resolve conflicts (if applicable)
4. Go back to `your_branch` with: `git checkout your_branch`
5. Merge `dev` branch into `your_brach` with: `git merge dev`
6. Resolve conflicts (if applicable)
7. Push changes to your fork with: `git push origin your_branch`
8. You can now create a pull request

Steps 1 -3 can be repeated for the `master` branch.

I'm still learning and practicing this process as well, so please let me know if you have any questions 😄

iEv0lv3 commented:
## Summary
First, this is a **great contribution**! I think it's a more elegant design than the standard `the_sort` method I wrote, and we should integrate it into the main code base `algorithm_sort.rb`. I finished reviewing and testing the code this afternoon, and there is only one minor issue:

* The `bubble_sort` method in it's current state does not consider word capitalization for sorting logic and output display. The way Ruby handles the alphabetical comparisons with the `>` and `<` operators is designed to always list capitalized words first. The current output of `bubble_sort` displays `['Congress', 'Republicans', 'Senate', 'alliance', 'constitution'...]`, with the capitalized words first. `the_sort` method I wrote has an example of one solution for this issue, but there are certainly more elegant ways to design the logic.
  * Implement functionality to consider capitalized words with `bubble_sort` and output display.

That's it; awesome work! 

## Next Steps
After the issue has been addressed I'd like to merge your code for the `AlgSort` class and associated `bubble_sort` method into the `dev` branch.
* For the next pull request it would be great if you could include your code for the `AlgSort` class in the `algorithm_sort.rb` file. For now please add it to the end of the file, just below where `sort` ends.

**Feature Requests**
* Integrating access to a `words` array (class attribute?) and the `ask_for_input` (class method?) method. This will enable a framework to really build out more features based around the `AlgSort` class.

**Discussion**
I agree with your suggestion that trying to separate the methods into logical classes would be beneficial. I haven't mapped out a solution to this yet, so any input is welcome, and I will take it into account while developing a plan.

* My initial thought is to have a class `Sort`,  and have class instances such as `bubble_sort`, `standard_sort`, `length_sort`, etc.
  * The question is what common attributes and methods would they share? Should we refactor all of the methods so they use the same `Sort` class template?
  * Or maybe create a class `DataSet`, and include the sorting methods as class methods?
  * Or store the sorting methods outside of the `DataSet` class, and access them as modules?

Thanks again for this contribution Steven; I really appreciate it! It's fun to  practice these things within our cohort. This is a great Bubble Sort, and I'm excited to get it included in the codebase!

iEv0lv3 commented on bubble_sort.rb:
```diff
@@ -0,0 +1,56 @@
+#
+#              Importantly, this works for an array of any data type. For strings, capitalized letters come first then lower case.
+#              Ruby treats comaprisons between letters by their ascii values. Meaning A = 65, B = 66 etc. and a = 97, b = 98 etc.
+#              More can be read here about how Ruby comapres strings: http://www.evc-cit.info/cit020/beginning-programming/chp_02/string_compare.html
```
I love the description and organization efforts!

For the purposes of trying to keep the code base readable as it gets bigger please remove this before creating a pull request. Thankfully GitHub tracks everything you contribute! 👍

alerrian replied:
- I will absolutely drop that in future pull requests, thanks for the input! You get used to a certain way of documenting stuff, and it's a habit of mine from college.
- Towards your initial comment regarding putting capitalized letters first. The reason this is like this, is because the default way to compare characters/strings is based on the ascii values of each character. Since the bubble will move the elements in to ascending order it will sort the smaller, capital letter values to the front of the array. I can probably find a way to alternate that if it makes more sense to keep the capitalized letter last?
- As far as feature requests, I will be adding those features to my code tomorrow, as I will have dedicated time for this. 

**Discussion points:**
My first instinct is to create a class that just includes each of the methods that we are writing for each *way* to sort. This would mean that we would basically need to just copy and paste the sorting methods in to the class and use dot-notation to access them. That said, I am a little iffy doing that as that makes actually using those methods a little un-intuitive. I'm going to do some research and play around with some options to see the best way we can clean it all up and make it all easier to use. I think your look makes a lot of sense but it will definitely be on us, as the programmers, to make sure we use meaningful instance naming. So, having a class that is called say, ``Sort`` would work nicely then include the sorting algs that were written so far (and the future ones) within that class. After that, calling those algs will come down to good naming convention. ``bubble_sort = Sort.new`` followed by ``bubble_sort.bubble_sort(name_of_the_array)`` will give us a lot more power outside of the class, while maintaining integrity of the class with basically no risk.   
   
So, I will respond to what you mentioned directly:
- So far we for sure have a template for a ``Sort`` class. The common methods will be all of the sorting algs we have so far. We might be able to work in a class that does ``WordProcessing`` or something to handle user input, but I think that may be better implemented right now strictly as a method, as it currently is.
- After looking some more at modules, I think that actually may be a really great way to implement these algorithms, since most of these algs, while taking in arrays mostly don't have a lot of the same things that will be within a regular class, though a class might need to implement them. So on second thought, I think maybe a ``Sort`` module may be the best way to implement these algs. That said, I think that keeping that module on a separate file is going to be the cleanest way to write it. Any other file that wants to access the module will need to use the `require_relative "name_of_file_without_the_extension"`
             `include NameOfTheModule` at the top of the file that wants to implement it.

alerrian replied:
Actually, I think I just realized what you meant with how it is sorting. Are you asking about getting the sort to sort each of the words on their alphabetical value, regardless of if it is capitalized or not?

iEv0lv3 replied:
Yes, your last comment is exactly what I intended to convey for sorting order. The goal is to have the order be alphabetical regardless of capitalized words being in the sort; using `.downcase` is great for this. There are additional steps to create the desired output of the sorted array in alphabetical order, including the capitalized words. The output display should also show the words in their original capitalized state. For example, using the `words` array: `"Senate"` would display in its correct dictionary alphabetical order, and also display as capitalized. View the output of running `sort(words)` for an example of this.

I'm new to giving feedback for code reviews and discussions, and will work to improve the clarity of my comments. I appreciate you practicing this process with me, and any feedback is welcome! Thanks for understanding while I learn 😄

alerrian replied:
Yeah, we're all learning! I just misunderstood what was being talked about. I should be able to make that change to the sort. I'll be working on some of this stuff hopefully today. Thanks for the feedback!

alerrian committed:
Add full alphabetical sort

alerrian commented:
The most recent commit adds the .downcase method to the sort and adds full functionality to arrays that hold strings/characters. Sadly, this method breaks the ability to sort numbers. Obviously, I could just rewrite the alg without downcase and call it bubble_sort_nums and it would be fine, but I think that is a ton of redundant code just to add functionality like that. What do you think?

EDIT: Actually, reviewing it, it might not turn out that bad if I continue with the class structure for collecting all of these algorithms. I only have to initialize the class once and then I can access the methods pretty much whenever. I think I will do what I said and just rename and write the same alg again, but for number use.

alerrian committed:
Change file names to better fit function. Add skeleton for new sorting alg

alerrian committed:
Add file skeleton for new class WordProcessing

alerrian committed:
Rename algorithm_sort to main, Move all sorts to sorting_class

alerrian commented:
I renamed a couple files to better fit what they were doing. I also moved all sorts to the sorting file, and renamed the algorithm_sort.rb file to main.rb. This means that any tests to be done or any time you want to make new objects it can be done in main. If you want to add a new sorting method just write it in the sorting_class.rb file then call it like normal.

alerrian commented:
Also, I apologize, I got all kinds of messed up with my commits so I committed several things at once rather than what I am sure is preferred to do just one thing in each commit.

alerrian committed:
Change WordProcessing to module

alerrian committed:
Add number sorting alg and some formatting

alerrian committed:
Add new number_sort method

iEv0lv3 reviewed (commented):
Hi Steven!

Apologies for the delay; I've been crushing the pre-work for the last few days and finishing up as many regular life tasks as I can before we start on Monday.

Top level: *really nice job* coding so many changes! You changed *355* lines of code, which is a lot of work...
![Alt Text](https://media.giphy.com/media/oYtVHSxngR3lC/giphy.gif)

Your coding skills are going to be on-point for day 1! I like how you started to build out a class and some class object methods as well, because that's the next big feature branch for us to develop. Solid use of reference files and a module! Some nice refactoring included as well.

I think the most reasonable course of action at this point is to break everything into manageable chunks, and then create pull requests built around a single idea or feature being implemented. That will make reviewing the changes and resolving merge conflicts much easier, more Agile focused, and will make sure no one change is so dramatic that it affects others working on the repository.

I'm still new to this, and I've been trying to research best practices in order to learn how I should maintain the repository, merge pull requests, complete code reviews, and give feedback. I appreciate you practicing and learning this process with me!
![Alt Text](https://media.giphy.com/media/wrzf9P70YWLJK/giphy.gif)

Here are a few resources I've been using to guide me through the pull request process with you:
[The anatomy of a perfect pull request](https://medium.com/@hugooodias/the-anatomy-of-a-perfect-pull-request-567382bb6067)
[The Art of Pull Requests](https://hackernoon.com/the-art-of-pull-requests-6f0f099850f9)

### Next Steps
* Your Bubble Sort is awesome and I want to get it merged as a feature!
* I'm going to close this pull request after you have a chance to review the comments, because we've covered a lot outside of the primary feature branch that will end up being merged.
* `git pull origin dev` into your local `dev` branch; it was updated today with Alexis's pull request.
* Create a new branch called `bubble_sort` from the updated dev branch. No need to delete your `SAA` branch, because your work will be a good guide for the next feature branch `classes_modules`.
* Modify the algorith_sort.rb file by adding your sort class that includes the `bubble_sort` method to the end of the file, and test that your code works.
* Create a new pull request to merge your `bubble_sort` branch into the `dev` branch.

Following the above process I should be able to instantly merge the `bubble_sort` branch into the `dev` branch because there should be 0 merge conflicts, and nothing to review with your code. Then we can start on the next feature branch from a working code base, which should allow us to implement, test, and commit small features more rapidly.
![Alt Text](https://media.giphy.com/media/du3J3cXyzhj75IOgvA/giphy.gif)

Let me know if you have any questions or comments, I'm doing the best I can for a noob 😏 Thanks again for practicing with me, and I'm looking forward to starting actual class with you on Monday!

iEv0lv3 commented:
Closing this pull request now that `bubble_sort` was merged into `dev` from a different branch. Please feel free to open pull requests to initiate code discussions , even if the code isn't ready! Issue and Feature requests are great for this as well. Thank you for the contributions!
